controlplane/internal/controller/agentcontrolplane_controller.go

```diff
@@ -199,6 +206,45 @@ func (r *AgentControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	return ctrl.Result{}, rerr
 }
 
+// Ensures dependencies are deleted before allowing the AgentControlPlane to be deleted
+// Deletes the ClusterDeployment (which deletes the AgentClusterInstall)
+// Machines, InfraMachines, and AgentBootstrapConfigs get auto-deleted when the ACP has a deletion timestamp - this deprovisions the BMH automatically
+// TODO: should we handle watching until all machines & agentbootstrapconfigs are deleted too?
+func (r *AgentControlPlaneReconciler) handleDeletion(ctx context.Context, acp *controlplanev1beta1.AgentControlPlane) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if controllerutil.ContainsFinalizer(acp, acpFinalizer) {
+		// Delete cluster deployment
+		if err := r.deleteClusterDeployment(ctx, acp.Status.ClusterDeploymentRef); err != nil && !apierrors.IsNotFound(err) {
+			log.Error(err, "failed deleting cluster deployment for ACP")
+			return ctrl.Result{}, err
+		}
+		log.Info("ACP's clusterdeployment is deleted")
+		acp.Status.ClusterDeploymentRef = nil
+
+		if controllerutil.RemoveFinalizer(acp, acpFinalizer) {
+			if err := r.Client.Update(ctx, acp); err != nil {
+				log.Error(err, "failed updating the ACP finalizer")
+				return ctrl.Result{}, err
+			}
+		}
+	}
+	log.Info("ACP doesn't contain finalizer, allow deletion")
```

Comment:
This will be shown even when finalizer has just been removed. Also this would probably be better as debug, WDYT?

Reply:
Modified it!